### PR TITLE
add settings for disable/enable LTR plugin

### PR DESCRIPTION
**Pull request by jngz-es**
#304

### CODE REVIEW

**Comment by nathancday** [comment]
Awesome, thanks for donating this! Give me some time to get this reviewed and I'll be back. Thanks again for building this out for the plugin.

**Comment by worleydl** [comment]
Will review this week.

**Comment by jngz-es** [comment]
> Will review this week.

Thanks.

**Comment by worleydl** [comment]
I have some edits I'm working on for this.  @nomoa what do you think of this concept as a whole?  I haven't seen enable/disable settings in other plugins I wonder if this is better served elsewhere somehow instead of having code scattered around to check the setting within the plugin.

@jngz-es is this a common pattern you've seen in other plugins?

File: src/main/java/com/o19s/es/explore/PostingsExplorerQuery.java [hunk 1]
```diff
@@ -78,6 +79,10 @@ public int hashCode() {
     @Override
     public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost)
             throws IOException {
+        if (!LTRSettings.isLTRPluginEnabled()) {
```
Comment (worleydl) [inline src/main/java/com/o19s/es/explore/PostingsExplorerQuery.java:82]: Possible edit is making this a "checkAndThrow" type of call as to not have the exception throw call be duplicated.

**Comment by jngz-es** [comment]
> I have some edits I'm working on for this. @nomoa what do you think of this concept as a whole? I haven't seen enable/disable settings in other plugins I wonder if this is better served elsewhere somehow instead of having code scattered around to check the setting within the plugin.
> 
> @jngz-es is this a common pattern you've seen in other plugins?

@worleydl https://github.com/opendistro-for-elasticsearch/anomaly-detection, https://github.com/opendistro-for-elasticsearch/k-NN.

File: src/main/java/com/o19s/es/explore/PostingsExplorerQuery.java (hunk 1 above)
Comment (jngz-es) [reply src/main/java/com/o19s/es/explore/PostingsExplorerQuery.java:82]: Good point.

**Comment by nomoa** [comment]
> I have some edits I'm working on for this. @nomoa what do you think of this concept as a whole? I haven't seen enable/disable settings in other plugins I wonder if this is better served elsewhere somehow instead of having code scattered around to check the setting within the plugin.
> 

I haven't seen this pattern either and have the same concerns, a central point of the plugin is the feature store so perhaps guarding on this condition could made there and would be a stronger guarantee that the plugin can't be used (except the couple queries we have that do rely on it LtrQuery/MatchExplorer).

**Comment by jngz-es** [comment]
@worleydl @nomoa I got your points. Do you have any suggestion on this feature? I mean like where to have a central control?

**Comment by nomoa** [comment]
@jngz-es sorry for the late reply.

I think adding the check in:
- com.o19s.es.ltr.feature.store.index.Caches#cacheLoad
- com.o19s.es.ltr.ranker.parser.LtrRankerParserFactory#getParser
will cover the most important and most costly operations of the plugin and will make the plugin completely useless.

If the goal is to disable absolutely everything (e.f. feature management) then you'll have to add the checks to:
- all rest endpoints in the package `com.o19s.es.ltr.rest`
- com.o19s.es.explore.ExplorerQueryBuilder#doToQuery (this is the sole query usable without the infrastructure of the plugin, i.e. it could be moved into a separate plugin).

With this I think all features of the plugin should be disabled.

**Pull request closed without merging**